pkg/new-kernel/patches-5.10.x/0022-siemens-ipc127-leds.patch

```diff
@@ -0,0 +1,253 @@
++ * GNU General Public License ("GPL") version 2 as published by the Free
++ * Software Foundation.
++ *
++ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
++ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
++ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
++ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
++ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
++ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
++ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
++ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
++ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
++ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
++ * POSSIBILITY OF SUCH DAMAGE.
++ */
++
++#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
++
++#include <linux/dmi.h>
++#include <linux/err.h>
++#include <linux/init.h>
++#include <linux/io.h>
++#include <linux/kernel.h>
++#include <linux/leds.h>
++#include <linux/module.h>
++#include <linux/platform_device.h>
++
++#define IPC127_FCH_ACPI_MMIO_BASE 0xD0C50500
++#define IPC127_FCH_GPIO_BASE      IPC127_FCH_ACPI_MMIO_BASE /* we should have a longer window instead, but see below */
++#define IPC127_LEDON              0x00 /* FIXME: we can't seem to be able to turn it off */
++#define IPC127_LEDOFF             0x01 /* fully -- so we go between RED and GREEN        */
++#define IPC127_IOSIZE             sizeof(u8)
++
++/* LED private data */
++struct ipc127_led_priv {
++	struct led_classdev cdev;
++	void __iomem *addr; /* for ioread/iowrite */
++};
++#define cdev_to_priv(c) container_of(c, struct ipc127_led_priv, cdev)
++
++static struct ipc127_led_pdata {
++	struct platform_device *pdev;
++	struct ipc127_led_priv *pled;
++	spinlock_t lock;
++} *leds;
++
++/* LED profile */
++/* based on https://cache.industry.siemens.com/dl/dl-media/673/109762673/att_975118/v3/ipc127e_operating_instructions/en-US/index.html#29951493a19e32f30da71d71fa3159d4 */
++static const struct ipc127_led_profile {
++	const char *name;
++	enum led_brightness brightness;
++	unsigned long offset; /* for devm_ioremap */
++} ipc127_led_profile[] = {
++	{ "ipc127:red:1",   LED_OFF,  IPC127_FCH_GPIO_BASE + 0x1A0 },
++	{ "ipc127:green:1", LED_OFF,  IPC127_FCH_GPIO_BASE + 0x1A8 },
++	{ "ipc127:red:2",   LED_OFF,  IPC127_FCH_GPIO_BASE + 0x1C8 },
++	{ "ipc127:green:2", LED_OFF,  IPC127_FCH_GPIO_BASE + 0x1D0 },
++	{ "ipc127:red:3",   LED_OFF,  IPC127_FCH_GPIO_BASE + 0x1E0 },
++	{ "ipc127:green:3", LED_OFF,  IPC127_FCH_GPIO_BASE + 0x198 },
++};
++
++static const struct dmi_system_id ipc127_led_dmi_table[] __initconst = {
++	{
++		.ident = "ipc127",
++		.matches = {
++			DMI_MATCH(DMI_SYS_VENDOR, "SIEMENS AG"),
++			DMI_MATCH(DMI_PRODUCT_NAME, "SIMATIC IPC127E")
```

Comment:
Here the product is 127E, but the name of the driver and kconfig is 127 without the 'E'. Do you think that will be confusing? (I don't know if there are other IPC127* models around.)

Reply:
The idea is to match to a family of IPC127* (I *think* -- but have no proof until I see a box -- that they are using letter to denote various additional component on the motherboard -- E seems to have TPM for example). Hence `.ident`. But then I'd like to only add to the `.matches` table after I get a chance to try it on a box. Given that this driver pokes in random memory areas -- I *REALLY* don't want it to activate accidentally.

That was the logic.

Reply:
OK